src/transforms/reduce/mod.rs

```diff
@@ -366,19 +381,30 @@ impl Reduce {
     }
 
     fn transform_one(&mut self, output: &mut Vec<Event>, event: Event) {
-        let (starts_here, event) = match &self.starts_when {
+        let (mut starts_here, event) = match &self.starts_when {
             Some(condition) => condition.check(event),
             None => (false, event),
         };
 
-        let (ends_here, event) = match &self.ends_when {
+        let (mut ends_here, event) = match &self.ends_when {
             Some(condition) => condition.check(event),
             None => (false, event),
         };
 
         let event = event.into_log();
         let discriminant = Discriminant::from_log_event(&event, &self.group_by);
 
+        if let Some(max_events) = self.max_events {
+            if max_events == 1 {
+                starts_here = true;
```

Comment:
If we're going to bother to special-case `max_events == 1`, we should just immediately write it to `output` here, and then return.

If we let the code fall through down to the `starts_when`/`ends_when` checks, we're wasting cycles checking the reducer merge state map (when we know we won't have anything, by definition) and, pursuant to the comment you left in the `max_event_1` test, depending on the `starts_when` path actually means the latest event only gets flushed when the next one comes in or the reducer flush expiration window is up.

If we just emit here, we solve all of these problems much more cleanly.

Reply:
Hmm I think even in the `max_events = 1` case the event should go through the merging mechanisms, so that things like `merge_strategies.field_name = "array"` get applied - short-circuiting the merge state map would pass events through unchanged, no? (The array merge strategy might be the only one affected by this). I agree it would be better to write these immediately though..

Reply:
Hmm, yes, that _is_ a fair point.

With that in mind, I think you should switch to setting `ends_with` instead of `starts_with` if `max_events == 1`, the way you sort of had it before.

The conditional below that checks if `ends_with` is true already handles generating a one-off reducer state with the given event if there was no entry in the reducer state map, which would address your concern here but still lead to immediate emission.

Reply:
Most of the merge strategies do end up being identity functions in the single-event case, so if array (or possible future non-identity strategies) is not in use that would be a potential optimization..though you might as well remove the `reduce` altogether for pass-through behavior like this.

Reply:
Agreed, the only trouble with `ends_when` was that the first event seen was handled differently (initialized in the `ReduceState` constructor). Some changes around that that and events now get flushed through immediately with `max_events=1`

Reply:
Ah yep, I see what you're saying. Your changes make sense to me. 👍🏻